### PR TITLE
Fix local builds that use cache or additional command flags

### PR DESCRIPTION
**Pull request by AndyTitu**
Related issue: https://github.com/1Password/shell-plugins/issues/90

### Why

The `Cache` and `CommandLine` field were left nil in the provision output response. This caused panics for both plugins using `Cache` and plugins using additional command flags.

Actually, the desired response of the RPC call should be the `ProvisionOutput` exactly as it was passed down from OP (as @Marton6 noted), but for a certain reason the fields of the pointer argument (`output`) of the call were nullified. (were brought to their zero values).

### What

this MR:
- adds a `ProvisionOutput` field to the `ProvisionCredentialRequest`., and uses it as response of the RPC call
- adds functionality for catching a panic from a locally built plugin and outputting it before the OP panic is reached, so that the OP panic doesn't mask the true cause of the problem.

### CODE REVIEW

File: sdk/rpc/server/server.go [hunk 1]
```diff
@@ -136,17 +137,18 @@ func (t *RPCServer) CredentialProvisionerDescription(req proto.ProvisionerID, re
 // interface. The call is forwarded to the Provision() function of the Provisioner of the credential identified by
 // req.CredentialID.
 func (t *RPCServer) CredentialProvisionerProvision(req proto.ProvisionCredentialRequest, resp *sdk.ProvisionOutput) error {
+	defer func() {
+		if err := recover(); err != nil {
+			resp.AddError(fmt.Errorf("your locally built plugin failed with the following panic: %s; and with stack trace: %s", err, string(debug.Stack())))
```
Comment (Marton6) [inline sdk/rpc/server/server.go:142]:
```suggestion
			resp.AddError(fmt.Errorf("your locally-built plugin panicked with the following error: %s; and with stack trace: %s", err, string(debug.Stack())))
```

**Review by Marton6 — CHANGES_REQUESTED** [review]
I have a small suggestion about the messaging around panics. Otherwise this MR looks good to me.

File: sdk/rpc/server/server.go (hunk 1 above)
Comment (jpcoenen) [inline sdk/rpc/server/server.go:142]: I really like that we're improving what happens on a panic. But could we extract this into a separate PR? I'd like to have a discussion about the consistency between calls and whether we want to use `resp.AddError()` for this. I think there is something to say for having a special type in the `proto` package to encode panics.

File: sdk/rpc/server/server.go [hunk 2]
```diff
@@ -136,17 +137,18 @@ func (t *RPCServer) CredentialProvisionerDescription(req proto.ProvisionerID, re
...
 	if err != nil {
 		return err
 	}
-	*resp = sdk.ProvisionOutput{
-		Environment: make(map[string]string),
-		CommandLine: nil,
-		Files:       make(map[string]sdk.OutputFile),
-		Diagnostics: sdk.Diagnostics{},
-	}
+	*resp = req.ProvisionOutput
```
Comment (jpcoenen) [inline sdk/rpc/server/server.go:149]: This is a breaking change as it sets a strict requirement on the `req.ProvisionOutput` field. 

For consideration: if we move the initialization to a shared function instead, we can achieve the same without having to break the API. Alternatively, we have to bump `proto.Version`.

Comment (jpcoenen) [reply sdk/rpc/server/server.go:149]: If we are changing the API, let's make sure to change it for all requests.

**Comment by hculea** [comment]
Let's also tackle `DeprovisionOutput` within this MR, shall we?

File: sdk/rpc/server/server.go (hunk 2 above)
Comment (hculea) [reply sdk/rpc/server/server.go:149]: Question here (just by looking at this): even if we move the initialisation to a shared function, we would still need to set a strict requirement for `req.CommandLine` field ([see Flo's adjustment](https://github.com/1Password/shell-plugins/commit/6bc2bd73ca63c50288c93c31ea9dfb2dda0ccd5e)) in order to pass the command line arguments, is this correct?

If this is the case, I think this prompts for bumping the `proto.Version` either way.

File: sdk/rpc/server/server.go (hunk 1 above)
Comment (AndyTitu) [reply sdk/rpc/server/server.go:142]: Sure thing, I'll extract this in another PR.

File: sdk/rpc/server/server.go (hunk 2 above)
Comment (jpcoenen) [reply sdk/rpc/server/server.go:149]: That's fair. I am not 100% happy with the fact that `req.CommandLine` (an output) is populated at the start of `Provision()`, but we can tackle that later and separately. 

The most important thing for now is that we make sure to merge this into the `protocol-v3` branch.

**Comment by jpcoenen** [comment]
> Let's also tackle `DeprovisionOutput` within this MR, shall we?

Yes, same for `Import()`.

File: sdk/rpc/server/server.go (hunk 2 above)
Comment (AndyTitu) [reply sdk/rpc/server/server.go:149]: This is indeed a protocol breaking change. I agree with Horia that there is no way around incrementing `port.Version`. The most elegant solution here (which would also avoid breaking protocol) would have been to use the `resp` parameter of the function directly as parameter for the `Provision` call here, but that pointer ends up pointing to a new struct when the rpc call is made. That is why I had to pass the output as a struct through the request structure.

**Comment by AndyTitu** [comment]
> Let's also tackle `DeprovisionOutput` within this MR, shall we?

> Yes, same for ImportOutput.

Done!

File: sdk/rpc/server/server.go (hunk 1 above)
Comment (AndyTitu) [reply sdk/rpc/server/server.go]: Thanks for the suggestion! I will apply this in the separate MR made especially for this panic catching.

**Comment by jpcoenen** [comment]
Changed the base branch to `protocol-v3`.

**Comment by jpcoenen** [comment]
I'm also going to rebase this branch onto `protocol-v3`. The current diff is a mess.

**Review by jpcoenen — APPROVED** [review]

File: sdk/rpc/server/server.go [hunk 3]
```diff
@@ -148,13 +149,9 @@ func (t *RPCServer) CredentialProvisionerProvision(req proto.ProvisionCredential
...
 	provisioner.Provision(context.Background(), req.ProvisionInput, resp)
+
```
Comment (hculea) [inline sdk/rpc/server/server.go:154]: tinny tiny nit: Let's remove this newline for consistency 😄

**Review by hculea — APPROVED** [review]

File: sdk/rpc/server/server.go (hunk 3 above)
Comment (hculea) [reply sdk/rpc/server/server.go:154]: If this is going to reset all the approvals, feel free to ignore this.

Comment (jpcoenen) [reply sdk/rpc/server/server.go:154]:
```suggestion
```

**Comment by AndyTitu** [comment]
For reference, Horia and I functionally tested this for both `mysql` and `aws` plugins.
